src/main/java/org/opensearch/knn/index/codec/transfer/VectorTransferByteToFloat.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.transfer;
+
+import org.opensearch.knn.index.codec.util.SerializationMode;
+import org.opensearch.knn.jni.JNICommons;
+
+import java.io.ByteArrayInputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Vector transfer for byte vectors by casting them as float vectors
+ */
+public class VectorTransferByteToFloat extends VectorTransfer {
+    private List<float[]> vectorList;
+
+    public VectorTransferByteToFloat(final long vectorsStreamingMemoryLimit) {
+        super(vectorsStreamingMemoryLimit);
+        vectorList = new ArrayList<>();
+    }
+
+    @Override
+    public void init(long totalLiveDocs) {
+        this.totalLiveDocs = totalLiveDocs;
+        vectorList.clear();
+    }
+
+    @Override
+    public void transfer(ByteArrayInputStream byteStream) {
+        final float[] vector = byteToFloatArray(byteStream);
+        dimension = vector.length;
+
+        if (vectorsPerTransfer == Integer.MIN_VALUE) {
+            vectorsPerTransfer = (dimension * Float.BYTES * totalLiveDocs) / vectorsStreamingMemoryLimit;
+            // This condition comes if vectorsStreamingMemoryLimit is higher than total number floats to transfer
+            // Doing this will reduce 1 extra trip to JNI layer.
+            if (vectorsPerTransfer == 0) {
+                vectorsPerTransfer = totalLiveDocs;
+            }
+        }
+
+        vectorList.add(vector);
+        if (vectorList.size() == vectorsPerTransfer) {
+            transfer();
+        }
+
+    }
+
+    @Override
+    public void close() {
+        transfer();
+    }
+
+    @Override
+    public SerializationMode getSerializationMode(ByteArrayInputStream byteStream) {
+        return SerializationMode.COLLECTION_OF_FLOATS;
+    }
+
+    private void transfer() {
+        vectorAddress = JNICommons.storeVectorData(vectorAddress, vectorList.toArray(new float[][] {}), totalLiveDocs * dimension);
+        vectorList.clear();
+    }
```

Comment:
If you are just going to transfer vectors are floats to JNI layer then the memory savings at the indexing layer will not be present because you are still transferring byte values as float point values. This is a big concern for me for enabling the byte vector support in faiss. and I can see these 2 paths here:

1. why we are not exposing a byte array based interfaces from faiss
2. Lets just transfer bytes to JNI layer and then on the JNI layer we convert byte vector to float vector one by one and send it faiss one by one for graph creation. This will ensure that memory is not spiking during indexing.

if 1 is not possible then we should definitely use 2 here otherwise benefits of byte vector for floats will not be same as that of Lucene.

Reply:
Got it. The Scalar quantizer in faiss accepts input as fp32 datatype. So, it might need a lot of changes and first approach may not be feasible. Will implement the second approach. Thanks!